### DISCUSSION TITLE
What is the difference between onLoadError, onLoadSucces and onSourceSuccess, onSourceError

### DISCUSSION THREAD

**Discussion by joseDaKing** [wojtekmaj/react-pdf#881] [Q&A] [answered]
What is the difference between the Document props onLoadError, onLoadSucces and onSourceSuccess, onSourceError? and when should use them ?

**Comment by wojtekmaj** [comment, marked as answer]
When `<Document />` is mounted and `file` prop is provided, React-PDF begins to load your file in two stages:
* source
* load

During source stage, document is retrieved from `file` prop.

`onSourceSuccess` is called when the document source is found and the file is *actually* about to begin loading or, somewhat controversially now that I think about it, when `file` is falsy and resolves to `null`, no file.

`onSourceError` is called when:
* Invalid dataURI is provided and `atob` or `unescape` fails when trying to decode binary data
* `file` is neither a string, nor PDFDataRangeTransport, nor ArrayBuffer, nor Blob (browser only), nor File (browser only), not an object.
* `file` is an object, but neither `url`, nor `data`, nor `range` property is found inside.

`onLoadProgress` will be called, 0-multiple times, when the file is in progress of being loaded (e.g. downloaded from the server).

`onLoadSuccess` is called when the document is finally loaded. This does NOT mean that something has been rendered yet - this is `<Page />`'s responsibility!

`onLoadError` is called when the document failed to load, e.g. when:
* React-PDF failed to download the file
* Corrupted PDF file was provided

In `<Page />`:

`onRenderSuccess` is called when the page is rendered.

`onRenderError` is called when the page failed to render.
